dcmgr/lib/dcmgr/endpoints/12.03/virtual_data_centers.rb

```diff
@@ -0,0 +1,75 @@
+# -*- coding: utf-8 -*-
+
+require 'dcmgr/endpoints/12.03/responses/virtual_data_center'
+require 'yaml'
+
+Dcmgr::Endpoints::V1203::CoreAPI.namespace '/virtual_data_centers' do
+  VDC_INSTANCE_TYPE = ['docker', 'openstack'].freeze
+  VDC_SPEC  = ['small', 'medium', 'large'].freeze
+
+  # Show list of virtual_data_centers
+  get do
+    ds = M::VirtualDataCenter.dataset
+    ds = datetime_range_params_filter(:created, ds)
+    ds = datetime_range_params_filter(:deleted, ds)
+
+    collection_respond_with(ds) do |paging_ds|
+      R::VirtualDataCenterCollection.new(paging_ds).generate
+    end
+  end
+
+  # Show detail of virtual_data_center
+  # param :id, string, requried
+  get '/:id' do
+    vdc = find_by_uuid(:VirtualDataCenter, params['id'])
+    raise E::UnknownVirtualDataCenter, params['id'] if vdc.nil?
+
+    respond_with(R::VirtualDataCenter.new(vdc).generate)
+  end
+
+  # Create virtual_data_center
+  # param :type, string, required
+  # param :spec: string, required
+  # param :spec_file: string, optional
+  post do
+    raise E::InvalidParameter, 'type' unless VDC_INSTANCE_TYPE.include? params['type']
+    raise E::InvalidParameter, 'spec' unless VDC_SPEC.include? params['spec']
```

Comment:
We are not checking for an invalid parameter but for missing required parameters. It should raise `:UndefinedRequiredParameter` instead.

https://github.com/axsh/wakame-vdc/blob/develop/dcmgr/lib/dcmgr/endpoints/errors.rb#L91